File: taurus-protect-sdk-typescript/src/models/webhook-call.ts
/**
 * Webhook call models for Taurus-PROTECT SDK.
 */

import type { CursorNavigationOptions, CursorPage } from './pagination';

/**
 * Webhook call status enum.
 */
export type WebhookCallStatus = 'SUCCESS' | 'FAILED' | 'PENDING';

/**
 * A webhook call record representing a single invocation of a configured webhook.
 *
 * A webhook call represents a single HTTP request made to a webhook URL
 * when an event occurs in Taurus-PROTECT. It includes information about
 * the payload sent, the delivery status, and retry attempts.
 */
export interface WebhookCall {
  /** Unique call identifier */
  readonly id?: string;
  /** The event ID that triggered this call */
  readonly eventId?: string;
  /** The webhook ID that was called */
  readonly webhookId?: string;
  /** The payload sent in the webhook call */
  readonly payload?: string;
  /** The status of the call (SUCCESS, FAILED, PENDING) */
  readonly status?: string;
  /** Status message with details about the call result */
  readonly statusMessage?: string;
  /** Number of delivery attempts */
  readonly attempts?: string;
  /** Last modification date */
  readonly updatedAt?: Date;
  /** Creation date */
  readonly createdAt?: Date;
}

/**
 * Options for listing webhook calls. A cursor list: `pageSize` 1-100 (default 20) and the
 * `cursor` of a previous page.
 */
export interface ListWebhookCallsOptions extends CursorNavigationOptions {
  /** Filter by event ID */
  eventId?: string;
  /** Filter by webhook ID */
  webhookId?: string;
  /** Filter by call status (SUCCESS, FAILED, PENDING) */
  status?: WebhookCallStatus;
  /** Sort order (ASC or DESC, default DESC) */
  sortOrder?: 'ASC' | 'DESC';
}

/**
 * A page of webhook calls.
 */
export interface WebhookCallResult {
  /** The webhook calls in this page */
  readonly calls: WebhookCall[];
  /** Cursor pagination: continue with `cursor: pagination.nextCursor` while `hasMore` */
  readonly pagination: CursorPage;
}

File: taurus-protect-sdk-go/pkg/protect/model/webhook_call.go
package model

import "time"

// WebhookCall represents a webhook call record.
type WebhookCall struct {
	// ID is the unique identifier for the webhook call.
	ID string `json:"id"`
	// EventID is the ID of the event that triggered this webhook call.
	EventID string `json:"event_id"`
	// WebhookID is the ID of the webhook that was called.
	WebhookID string `json:"webhook_id"`
	// Payload is the JSON payload sent to the webhook.
	Payload string `json:"payload,omitempty"`
	// Status is the status of the webhook call (e.g., "SUCCESS", "FAILED", "PENDING").
	Status string `json:"status"`
	// StatusMessage provides additional details about the status.
	StatusMessage string `json:"status_message,omitempty"`
	// Attempts is the number of attempts made to deliver this webhook.
	Attempts int64 `json:"attempts"`
	// CreatedAt is when the webhook call was created.
	CreatedAt time.Time `json:"created_at"`
	// UpdatedAt is when the webhook call was last updated.
	UpdatedAt time.Time `json:"updated_at"`
}

// ListWebhookCallsOptions contains options for listing webhook calls.
type ListWebhookCallsOptions struct {
	// EventID filters by the event ID that triggered the webhook calls.
	EventID string
	// WebhookID filters by the webhook ID.
	WebhookID string
	// Status filters by the webhook call status (e.g., "SUCCESS", "FAILED", "PENDING").
	Status string
	// CurrentPage is the base64-encoded cursor for the current page.
	CurrentPage string
	// PageRequest indicates which page to request (FIRST, PREVIOUS, NEXT, LAST).
	PageRequest string
	// PageSize is the number of items per page.
	PageSize int64
	// SortOrder specifies the sort order (ASC or DESC). Default is DESC.
	SortOrder string
}

// ListWebhookCallsResult contains the result of listing webhook calls.
type ListWebhookCallsResult struct {
	// WebhookCalls is the list of webhook call records.
	WebhookCalls []*WebhookCall `json:"webhook_calls"`
	// CurrentPage is the base64-encoded cursor for the current page.
	CurrentPage string `json:"current_page,omitempty"`
	// HasPrevious indicates if there is a previous page.
	HasPrevious bool `json:"has_previous"`
	// HasNext indicates if there is a next page.
	HasNext bool `json:"has_next"`
}
